### PR TITLE
Replace with using stride offset to load descriptor

### PR DESCRIPTION
**Pull request by crystalJinamd**
In order to accordance with combined image sampler type in spec, replace
with single descriptor stride offset of per-plane SRD to address a image
desripotr rather than a fixed image descriptor size.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Comment by trenouf** [comment]
Thanks for pushing this Crystal. But I still don't understand what srdStride is for. In what situation does it need to be different to stride? What problem is it solving? Is this enabling some future functionality?

**Comment by crystalJinamd** [comment]
retest this please

**Review by nhaehnle — CHANGES_REQUESTED** [review]
I'm inclined to say definitely No to this change.

The way that YCbCr is spec'd is absolutely horrible. While we do have to support it, we really should do so in a way that minimizes the blast radius. The ResourceNode stuff is opaque and confusing enough as-is, let's try to avoid making it more so.

To that end, section 13.1 (Sampler Y′CBCR conversion) of the Vulkan 1.3.234 spec says:

> To create a sampler Y′CBCR conversion, the [samplerYcbcrConversion](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#features-samplerYcbcrConversion) feature must be enabled. Conversion must be fixed at pipeline creation time, through use of a combined image sampler with an immutable sampler in VkDescriptorSetLayoutBinding.

In other words, the images we're loading **always** come from a binding with VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER type. Which means that descriptor set layout calculation always flows down the same path, which means that the stride with which image plane descriptors are stored is always the same.

There simply doesn't seem to be a need to make this dynamic in the compiler.

**Comment by qiaojbao** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1ccf428d9a5ea47f5be0f30de7c7432d54916b75
<details><summary>CTS tests (Failed: 1/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38904/70473 (55.2%)</li>
<li><details><summary>Failed: 1/70473 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.ssbo.layout.unsized_nested_struct_array.per_block_buffer.std140_instance_array_comp_access
Stack trace: Flake
```
</p></details></li>

<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by qiaojbao** [comment]
retest this please

File: lgc/builder/BuilderImpl.h [hunk 1]
```diff
@@ -292,7 +292,7 @@ class DescBuilder : virtual public BuilderImplBase {
 
   // Create a get of the stride (in bytes) of a descriptor.
   llvm::Value *CreateGetDescStride(ResourceNodeType concreteType, ResourceNodeType abstractType, unsigned descSet,
-                                   unsigned binding, const llvm::Twine &instName) override final;
+                                   unsigned binding, unsigned getSrdStride, const llvm::Twine &instName) override final;
```
[the remark this answers is not in the document]
Comment (crystalJinamd) [reply lgc/builder/BuilderImpl.h]: The design is required for new extension, and has been under discussion offline.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 2]
```diff
@@ -2596,9 +2596,14 @@ Value *SPIRVToLLVM::loadImageSampler(Type *elementTy, Value *base) {
...
-    if (!m_convertingSamplers.empty()) {
+    if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() >= 3) {
+      Value *srdStride = getBuilder()->CreateExtractValue(base, 2);
+      Type *ptrTy = ptr->getType();
+
       for (unsigned planeIdx = 1; planeIdx != arrayTy->getNumElements(); ++planeIdx) {
-        ptr = getBuilder()->CreateGEP(elementTy, ptr, getBuilder()->getInt32(1));
+        ptr = getBuilder()->CreatePtrToInt(ptr, getBuilder()->getInt64Ty());
+        ptr = getBuilder()->CreateAdd(ptr, getBuilder()->CreateZExt(srdStride, getBuilder()->getInt64Ty()));
+        ptr = getBuilder()->CreateIntToPtr(ptr, ptrTy);
```
[the remark this answers is not in the document]
Comment (crystalJinamd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: Fixed in the latest commit, please help to review, thanks !

**Comment by amdvlk-admin** [comment]
### Test summary for commit de8f1b3316a505cf497537c7e48ded83f92a15a0
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by qiaojbao** [comment]
retest this please

**Comment by qiaojbao** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 858921938c7026bdcdb480a1646f6585211a8601
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
Note, there is an internal discussion ongoing at the moment. This was triggered by VK_EXT_descriptor_buffer, which has now been published, but IMHO we should just implement the YCbCr aspect of that extension differently.

**Comment by qiaojbao** [comment]
retest this please

**Comment by qiaojbao** [comment]
retest this please

**Comment by qiaojbao** [comment]
### Test summary for commit 858921938c7026bdcdb480a1646f6585211a8601
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -2703,12 +2709,38 @@ Value *SPIRVToLLVM::transImagePointer(SPIRVValue *spvImagePtr) {
 Value *SPIRVToLLVM::getDescPointerAndStride(ResourceNodeType resType, unsigned descriptorSet, unsigned binding,
                                             ResourceNodeType searchType) {
   if (resType != ResourceNodeType::DescriptorSampler) {
-    // Image/f-mask/texel buffer, where a pointer is represented by a struct {pointer,stride}.
+    // f-mask/texel buffer, where a pointer is represented by a struct {pointer,stride}.
     Value *descPtr = getBuilder()->CreateGetDescPtr(resType, searchType, descriptorSet, binding);
     Value *descStride = getBuilder()->CreateGetDescStride(resType, searchType, descriptorSet, binding);
     descPtr = getBuilder()->CreateInsertValue(
-        UndefValue::get(StructType::get(*m_context, {descPtr->getType(), descStride->getType()})), descPtr, 0);
+        UndefValue::get(
```
Comment (nhaehnle) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:2716]: While we're changing this code, use `PoisonValue` here (LLVM transition from undef to poison).

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 4]
```diff
@@ -3154,7 +3186,11 @@ Value *SPIRVToLLVM::indexDescPtr(Type *elementTy, Value *base, Value *index) {
   // A sampler pointer is represented by a {pointer,stride,convertingSamplerIdx} struct. If the converting sampler
   // index is non-zero (i.e. it is actually a converting sampler), we also want to modify that index. That can only
   // happen if there are any converting samplers at all.
-  if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() >= 3) {
+  Type *pointerTy = base->getType()->getStructElementType(0);
+  unsigned vecCount = (pointerTy->getNumContainedTypes() != 0)
+                          ? cast<FixedVectorType>(pointerTy->getPointerElementType())->getNumElements()
+                          : 0;
+  if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() >= 3 && vecCount == 4) {
```
Comment (nhaehnle) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:3193]: I don't understand why this change is necessary. Also, I don't think it will work with opaque pointers.

**Review by nhaehnle — COMMENTED** [review]
Thank you, this looks okay, just some inline comments left.

The commit message could benefit from some explanation that we're doing this for compatibility with SC, which has to layout YCbCr descriptors with 12 dwords stride per plane for VK_EXT_descriptor_buffer support.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp (hunk 4 above)
Comment (crystalJinamd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:3193]: There would be overlapped when translate load sampler and image descriptor. These codes are used to distinguish sampler and image descriptors loading. 

The commit could be pass Docker testings which include "amdllpc -enable-opaque-pointers=true"

Comment (nhaehnle) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:3193]: That the tests pass probably just means that there's a lack of proper test coverage for this.

Really, this code pattern simply isn't okay. In an opaque pointer world, you get `getNumContainedTypes() == 0`, which means you get `vecCount == 0`, which means this code never executes. That can't be right.

Also, I don't understand what you mean by overlap.

Comment (crystalJinamd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:3193]: The content here is used to load sampler descriptor. If only check with "if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() >= 3)", it would load image descriptor with sampler operations. Such as:

  %7 = call i32 (...) @lgc.create.get.desc.stride.i32(i32 1, i32 1, i32 1, i32 0)
  %8 = insertvalue { <8 x i32> addrspace(4)*, i32, i32 } poison, <8 x i32> addrspace(4)* %6, 0
  %9 = insertvalue { <8 x i32> addrspace(4)*, i32, i32 } %8, i32 %7, 1
  %10 = udiv i32 %7, 3
  %11 = insertvalue { <8 x i32> addrspace(4)*, i32, i32 } %9, i32 %10, 2
  %12 = insertvalue { { <8 x i32> addrspace(4)*, i32, i32 }, { <4 x i32> addrspace(4)*, i32, i32 } } undef, { <8 x i32> addrspace(4)*, i32, i32 } %11, 0
  %13 = insertvalue { { <8 x i32> addrspace(4)*, i32, i32 }, { <4 x i32> addrspace(4)*, i32, i32 } } %12, { <4 x i32> addrspace(4)*, i32, i32 } { <4 x i32> addrspace(4)* null, i32 0, i32 1 }, 1
  %14 = extractvalue { { <8 x i32> addrspace(4)*, i32, i32 }, { <4 x i32> addrspace(4)*, i32, i32 } } %13, 0
  %15 = extractvalue { { <8 x i32> addrspace(4)*, i32, i32 }, { <4 x i32> addrspace(4)*, i32, i32 } } %13, 1
  **%16 = extractvalue { <8 x i32> addrspace(4)*, i32, i32 } %14, 2
  %17 = add i32 %16, 0
  %18 = icmp ne i32 %16, 0
  %19 = select i1 %18, i32 %17, i32 0
  %20 = insertvalue { <8 x i32> addrspace(4)*, i32, i32 } %14, i32 %19, 2**
  %21 = extractvalue { <8 x i32> addrspace(4)*, i32, i32 } %20, 0
  %22 = extractvalue { <8 x i32> addrspace(4)*, i32, i32 } %20, 1

And then it would go wrong into bold codes when handle following codes of SPIRVToLLVM::loadImageSampler(). Actually I'm not sure if it is a LLVM optimization latent issue.

    if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() >= 3) {
      Value *descSrdStride = getBuilder()->CreateExtractValue(base, 2);
      Type *ptrTy = ptr->getType();

      for (unsigned planeIdx = 1; planeIdx != arrayTy->getNumElements(); ++planeIdx) {
        ptr = getBuilder()->CreateBitCast(
            ptr, getBuilder()->getInt8Ty()->getPointerTo(ptr->getType()->getPointerAddressSpace()));
        ptr = getBuilder()->CreateGEP(getBuilder()->getInt8Ty(), ptr, descSrdStride);
        ptr = getBuilder()->CreateBitCast(ptr, ptrTy);
        oneVal = getBuilder()->CreateLoad(elementTy, ptr);
        result = getBuilder()->CreateInsertValue(result, oneVal, planeIdx);
      }
    }

**Comment by amdvlk-admin** [comment]
### Test summary for commit cffebe8090f4e19fdc43bddc1881017bb92cd529
<details><summary>CTS tests (Failed: 0/207274)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70473 (55.2%)</li>
<li>Failed: 0/70473 (0.0%)</li>
<li>Not Supported: 31568/70473 (44.8%)</li><li>Warnings: 0/70473 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70481 (50.7%)</li>
<li>Failed: 0/70481 (0.0%)</li>
<li>Not Supported: 34731/70481 (49.3%)</li><li>Warnings: 0/70481 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66320 (50.5%)</li>
<li>Failed: 0/66320 (0.0%)</li>
<li>Not Supported: 32844/66320 (49.5%)</li><li>Warnings: 0/66320 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp (hunk 4 above)
Comment (linqun) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:3193]: an alternate solution is adding two members in resource node.  {resource_ptr, stride, plane_stride, is_res}, then we only need change this condition to 
if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() == 3)

**Comment by amdvlk-admin** [comment]
### Test summary for commit 186e0c2b68c753e1e41f0d35c3d355f79698d1e0
<details><summary>CTS tests (Failed: 0/207271)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi2x</summary><ul>

<li>Passed: 38905/70472 (55.2%)</li>
<li>Failed: 0/70472 (0.0%)</li>
<li>Not Supported: 31567/70472 (44.8%)</li><li>Warnings: 0/70472 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 35750/70480 (50.7%)</li>
<li>Failed: 0/70480 (0.0%)</li>
<li>Not Supported: 34730/70480 (49.3%)</li><li>Warnings: 0/70480 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx9</summary><ul>

<li>Passed: 33476/66319 (50.5%)</li>
<li>Failed: 0/66319 (0.0%)</li>
<li>Not Supported: 32843/66319 (49.5%)</li><li>Warnings: 0/66319 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp (hunk 4 above)
Comment (crystalJinamd) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: Fixed in the latest commit, please help to review, thanks !

**Review by nhaehnle — APPROVED** [review]
Thank you!
